db/db_impl/db_impl.cc

```diff
@@ -1569,7 +1570,25 @@ Status DBImpl::ApplyWALToManifest(VersionEdit* synced_wals) {
   return status;
 }
 
-Status DBImpl::LockWAL() {
+Status DBImpl::LockWAL(std::vector<uint64_t>* required_by_manifest) {
+  Status deletions_disabled = DisableFileDeletions();
+  if (!deletions_disabled.ok()) {
+    return deletions_disabled;
+  }
+  {
+    InstrumentedMutexLock l(&mutex_);
+    while (pending_purge_obsolete_files_ > 0 || bg_purge_scheduled_ > 0) {
+      bg_cv_.Wait();
+    }
+    if (required_by_manifest) {
+      // Record tracked WALs as a (minimum) cross-check for directory scan
+      const auto& manifest_wals = versions_->GetWalSet().GetWals();
+      required_by_manifest->reserve(manifest_wals.size());
+      for (const auto& wal : manifest_wals) {
+        required_by_manifest->push_back(wal.first);
+      }
+    }
+  }
   log_write_mutex_.Lock();
```

Comment:
This doesn't look like it'd block writes like the original PR (#5146) claimed. Does it? I think explicitly blocking writes (like a write stop) may work better and be simpler.

I also fail to see how the backup worked before without calling `DisableFileDeletions()`. Did it work?

Reply:
The #5146 PR claimed that: "Walk through the storage engines and lock writes on each engine. For
InnoDB, redo log is locked. For MyRocks, WAL should be locked."
My original impression is that WAL is locked, thus any writes with WAL will be blocked after this call.

Reply:
Furthermore, I thought introducing write stop is undesirable for users, but maybe that's over-optimizing.

Reply:
> My original impression is that WAL is locked, thus any writes with WAL will be blocked after this call.

OK, I didn't have that impression by looking at how `log_write_mutex_` is used. I will test it then since my concern isn't shared.

Reply:
Taking another look at https://github.com/facebook/rocksdb/blob/7.9.fb/db/db_impl/db_impl_write.cc#L1213. In fact, all threads calling `PreprocessWrite()` may be blocked.

Reply:
It did not work originally but #7516 fixed it, probably without noticing. I still think we should re-think the approach as it feels unstable and unintuitive.

Reply:
I am also hesitant about exposing `log_write_mutex_` to application.
Maybe similar to what innodb does, we don't need keep the log_write_mutex locked as long as we disable file deletion.